internal/storage/dropbox/dropbox.go

```diff
@@ -0,0 +1,182 @@
+package dropbox
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox"
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox/files"
+	"github.com/offen/docker-volume-backup/internal/storage"
+)
+
+type dropboxStorage struct {
+	*storage.StorageBackend
+	client files.Client
+}
+
+// Config allows to configure a Dropbox storage backend.
+type Config struct {
+	Token      string
+	RemotePath string
+}
+
+// NewStorageBackend creates and initializes a new Dropbox storage backend.
+func NewStorageBackend(opts Config, logFunc storage.Log) (storage.Backend, error) {
+	if opts.Token == "" {
+		return nil, errors.New("NewStorageBackend: No Dropbox token has been provided")
+	} else {
+		config := dropbox.Config{
+			Token: opts.Token,
+		}
+
+		client := files.New(config)
+
+		return &dropboxStorage{
+			StorageBackend: &storage.StorageBackend{
+				DestinationPath: opts.RemotePath,
+				Log:             logFunc,
+			},
+			client: client,
+		}, nil
+	}
+}
+
+// Name returns the name of the storage backend
+func (b *dropboxStorage) Name() string {
+	return "Dropbox"
+}
+
+// Copy copies the given file to the WebDav storage backend.
+func (b *dropboxStorage) Copy(file string) error {
+	_, name := path.Split(file)
+
+	folderArg := files.NewCreateFolderArg(b.DestinationPath)
+	if _, err := b.client.CreateFolderV2(folderArg); err != nil {
+		if err.(files.CreateFolderV2APIError).EndpointError.Path.Tag == files.WriteErrorConflict {
+			b.Log(storage.LogLevelInfo, b.Name(), "Destination path '%s' already exists in Dropbox, no new directory required.", b.DestinationPath)
+		} else {
+			return fmt.Errorf("(*dropboxStorage).Copy: Error creating directory '%s' in Dropbox: %w", b.DestinationPath, err)
+		}
+	}
+
+	r, err := os.Open(file)
+	if err != nil {
+		return fmt.Errorf("(*dropboxStorage).Copy: Error opening the file to be uploaded: %w", err)
+	}
+	defer r.Close()
+
+	// Start new upload session and get session id
+
+	b.Log(storage.LogLevelInfo, b.Name(), "Starting upload session for backup '%s' to Dropbox at path '%s'.", file, b.DestinationPath)
+
+	var sessionId string
+	uploadSessionStartArg := files.NewUploadSessionStartArg()
+	uploadSessionStartArg.SessionType = &files.UploadSessionType{Tagged: dropbox.Tagged{Tag: files.UploadSessionTypeConcurrent}}
+	if res, err := b.client.UploadSessionStart(uploadSessionStartArg, nil); err != nil {
+		return fmt.Errorf("(*dropboxStorage).Copy: Error starting the upload session: %w", err)
+	} else {
+		sessionId = res.SessionId
+	}
+
+	// Send the file in 148MB chunks (Dropbox API limit is 150MB, concurrent upload requires a multiple of 4MB though)
+	// Last append can be any size <= 150MB with Close=True
+
+	const chunkSize = 148 * 1024 * 1024 // 148MB
+	var offset uint64 = 0
+
+	for {
```

Comment:
Is Dropbox capable of handling the upload of multiple chunks in parallel? In case yes, should we upload chunks here concurrently instead of serially?

Reply:
> Using one upload stream, I could saturate my ~45 MBit/s upload to Dropbox.

I would assume most times this tool is running on VPSes or similar with decent connections, so I'd rather optimize for faster connections, but then again the file size limit itself will prevent this from becoming a real bottleneck in either scenario.

Reply:
Seems like 140-160 MBit/s per stream is possible. In my case, 4 concurrent streams resulted in 666 MBit/s and 6 and above in 1000 MBit/s. Maybe the uplink of my server to Dropbox was the limit or they specifically limit it to 1 GBit/s... setting the default to 6 then.